src/validation/rules/unique_variable_names.rs

```diff
@@ -0,0 +1,94 @@
+use std::collections::HashMap;
+
+use super::ValidationRule;
+use crate::static_graphql::query::*;
+use crate::validation::utils::{ValidationError, ValidationErrorContext};
+use crate::{
+    ast::{ext::AstWithVariables, QueryVisitor},
+    validation::utils::ValidationContext,
+};
+
+/// Unique variable names
+///
+/// A GraphQL operation is only valid if all its variables are uniquely named.
+///
+/// See https://spec.graphql.org/draft/#sec-Variable-Uniqueness
+pub struct UniqueVariableNames;
+
+struct UniqueVariableNamesHelper<'a> {
+    error_context: ValidationErrorContext<'a>,
+}
+
+impl<'a> UniqueVariableNamesHelper<'a> {
+    fn new(validation_context: &'a ValidationContext<'a>) -> Self {
+        UniqueVariableNamesHelper {
+            error_context: ValidationErrorContext::new(validation_context),
+        }
+    }
+}
+
+impl<'a> QueryVisitor<UniqueVariableNamesHelper<'a>> for UniqueVariableNames {
+    fn leave_operation_definition(
+        &self,
+        node: &OperationDefinition,
+        visitor_context: &mut UniqueVariableNamesHelper<'a>,
+    ) {
+        let variables = node.get_variables();
+
+        let mut seen_variables = HashMap::new();
```

Comment:
Prefer to use `HashSet` for this kind of indicator.
Also, prefer to define it more strict: `HashSet::<String>::new()`